prov/cxi/src/cxip_mr.c

```diff
@@ -1153,6 +1454,23 @@ int cxip_mr_enable(struct cxip_mr *mr)
 	}
 	mr->optimized = cxip_generic_is_mr_key_opt(mr->key);
 
+	/* Force standard (unoptimized) MRs for RMA events (fi_writedata support).
+	 * Optimized MRs do not support event generation or header_data delivery
+	 * in target events. Therefore, creation of an optimized MR must be
+	 * forced to a standard MR if:
+	 * 1. RMA events are requested (mr->rma_events), OR
+	 * 2. Writedata is enabled and domain rma_cq_data_size is non-zero
+	 */
+	if (mr->rma_events ||
```

Comment:
I think you should constrain this check to: if (mr->domain->is_prov_key && (mr->rma_events || mr->domain->rma_cq_data_size)). Setting mr->optimized directly for a client key will not work as cxip_generic_is_mr_key_opt() will still return that it is an optimized MR key. Use of client keys should keep inject_writedata/writedata from being enabled and the source using the client key will know to use unrestricted packets due to FI_RMA_EVENTS.